### PR TITLE
Fix gcc 4.1 build (#1035)

### PR DESCRIPTION
**Pull request by matthauck**

### CODE REVIEW

**Comment by matthauck** [comment]
This should fix #1035

**Comment by matthauck** [comment]
_ping_ 

Is there anything missing here? Any changes requested?

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by TeBoring** [comment]
Can you rebase this pull request on the current master?

**Comment by matthauck** [comment]
Just finished doing that and making sure it still built. :)

**Comment by TeBoring** [comment]
ok to test.

**Comment by TeBoring** [comment]
LGTM

File: src/google/protobuf/map.h [hunk 1]
```diff
@@ -1726,12 +1726,19 @@ struct hash<google::protobuf::MapKey> {
...
       case google::protobuf::FieldDescriptor::CPPTYPE_INT64:
         return hash< ::google::protobuf::int64>()(map_key.GetInt64Value());
-      case google::protobuf::FieldDescriptor::CPPTYPE_INT32:
-        return hash< ::google::protobuf::int32>()(map_key.GetInt32Value());
       case google::protobuf::FieldDescriptor::CPPTYPE_UINT64:
         return hash< ::google::protobuf::uint64>()(map_key.GetUInt64Value());
+#else
+      case google::protobuf::FieldDescriptor::CPPTYPE_INT64:
+      case google::protobuf::FieldDescriptor::CPPTYPE_UINT64:
+        GOOGLE_LOG(FATAL) << "Unsupported on this platform.";
```
Comment (xfxyjwf) [inline src/google/protobuf/map.h]: This is not the right use of LOG(FATAL). Protobuf library should not crash an user's program given a valid input. If the platform can not be supported, it should be reported at compile time. Though, in this case it seems easy enough to write a custom hash function if it's not already available in the platform headers. @matthauck @TeBoring Do you plan to actually implement this hash for int64? I saw this change while working on integrating github changes to our internal repo and I'm going to leave this change out.

Comment (matthauck) [inline src/google/protobuf/map.h]: Oh, whoops, sorry. 😞 

The thought did not cross my mind to implement this on these platforms. I'd be open to helping out if you could give me some pointers on where to start.

Comment (TeBoring) [inline src/google/protobuf/map.h]: I guess just an empty implementation would be enough. It's better to trigger compile error if the hash for int64 is used.

Comment (TeBoring) [inline src/google/protobuf/map.h]: https://github.com/google/protobuf/blob/master/src/google/protobuf/stubs/hash.h#L184
Here is an example. For int64, how about just trimming?

Comment (matthauck) [inline src/google/protobuf/map.h]: Okay, opened up two PR's to fix this missing hash: #2906 and #2907. Take your pick which fix you like better. :)

**Pull request merged**